x-pack/plugin/geo/src/main/java/org/elasticsearch/xpack/geo/GeoLineAggregator.java

```diff
@@ -0,0 +1,131 @@
+        super(name, context, parent, pipelineAggregators, metaData);
+        this.valuesSources = valuesSources;
+        if (valuesSources != null) {
+            positions = context.bigArrays().newObjectArray(1);
+        }
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        if (valuesSources != null && valuesSources.needsScores()) {
+            return ScoreMode.COMPLETE;
+        }
+        return super.scoreMode();
+    }
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
+                                                final LeafBucketCollector sub) throws IOException {
+        if (valuesSources == null) {
+            return LeafBucketCollector.NO_OP_COLLECTOR;
+        }
+        final BigArrays bigArrays = context.bigArrays();
+        MultiGeoPointValues docGeoPointValues = valuesSources.getGeoPointField(GEO_POINT_FIELD.getPreferredName(), ctx);
+        SortedNumericDoubleValues docSortValues = valuesSources.getNumericField(SORT_FIELD.getPreferredName(), ctx);
+
+        return new LeafBucketCollectorBase(sub, docGeoPointValues) {
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                positions = bigArrays.grow(positions, bucket + 1);
+                if (docGeoPointValues.advanceExact(doc) && docSortValues.advanceExact(doc)) {
+                    if (docSortValues.docValueCount() > 1) {
+                        throw new AggregationExecutionException("Encountered more than one sort value for a " +
+                            "single document. Use a script to combine multiple sort-values-per-doc into a single value.");
+                    }
+                    if (docGeoPointValues.docValueCount() > 1) {
+                        throw new AggregationExecutionException("Encountered more than one geo_point value for a " +
+                            "single document. Use a script to combine multiple geo_point-values-per-doc into a single value.");
+                    }
+
+                    // There should always be one weight if advanceExact lands us here, either
+                    // a real weight or a `missing` weight
+                    assert docSortValues.docValueCount() == 1;
+                    assert docGeoPointValues.docValueCount() == 1;
+                    final double sort = docSortValues.nextValue();
+                    final GeoPoint point = docGeoPointValues.nextValue();
+                    ArrayList<InternalGeoLine.Position> bucketLine = positions.get(bucket);
+                    if (bucketLine == null) {
+                        bucketLine = new ArrayList<>(1000);
+                    } else if (bucketLine.size() >= 10000) {
+                        throw new AggregationExecutionException("Cannot aggregate a line with more than 10000 points");
+                    }
+                    bucketLine.add(new InternalGeoLine.Position(
```

Comment:
Was noodling over this and it'd be nice to avoid instantiating all these `Position` objects.  E.g. a dozen buckets with 10k points generates a lot of objects.

If we don't need the sort value to be a `double`, I'm wondering if perhaps we could use `PackedLongValues`, `PackedInts`, etc and encode the three values consecutively in the data-structure.  That'd avoid the object allocation, and gain the packing ability for maybe better compression.  

I suspect these won't pack well though, since I bet the lat/lon points requirea large number of bits per value.  But we could do similar with just a plain `long[]` and that'd also avoid the object allocations

The downside is that we'd have to allocate the whole block of memory up-front.  But I'm guessing most lines will be more-filled than less-filled so that's probably an ok trade-off.